src/MongoDB.EntityFrameworkCore/Serializers/SerializationHelper.cs

```diff
@@ -106,74 +107,55 @@ internal static BsonSerializationInfo GetPropertySerializationInfo(IReadOnlyProp
-            var t when t == typeof(short) => new Int16Serializer(),
-            var t when t == typeof(int) => Int32Serializer.Instance,
-            var t when t == typeof(long) => Int64Serializer.Instance,
-            var t when t == typeof(ObjectId) => ObjectIdSerializer.Instance,
-            var t when t == typeof(TimeSpan) => new TimeSpanSerializer(),
-            var t when t == typeof(sbyte) => new SByteSerializer(),
-            var t when t == typeof(float) => new SingleSerializer(),
-            var t when t == typeof(string) => new StringSerializer(),
-            var t when t == typeof(ushort) => new UInt16Serializer(),
-            var t when t == typeof(uint) => new UInt32Serializer(),
-            var t when t == typeof(ulong) => new UInt64Serializer(),
-            var t when t == typeof(Decimal128) => new Decimal128Serializer(),
-            var t when t.IsEnum => EnumSerializer.Create(t),
-            {IsArray: true} a => CreateArraySerializer(a.GetElementType()),
-            {IsGenericType: true} l => CreateListSerializer(l.TryGetItemType(typeof(IEnumerable<>))),
-            _ => throw new NotSupportedException($"Cannot resolve Serializer for '{type.FullName}' type."),
+            _ when type == typeof(bool) => BooleanSerializer.Instance,
+            _ when type == typeof(byte) => new ByteSerializer(),
+            _ when type == typeof(char) => new CharSerializer(),
+            _ when type == typeof(DateTime) => CreateDateTimeSerializer(property),
+            _ when type == typeof(DateTimeOffset) => new DateTimeOffsetSerializer(),
+            _ when type == typeof(decimal) => new DecimalSerializer(),
+            _ when type == typeof(double) => DoubleSerializer.Instance,
+            _ when type == typeof(Guid) => new GuidSerializer(),
+            _ when type == typeof(short) => new Int16Serializer(),
+            _ when type == typeof(int) => Int32Serializer.Instance,
+            _ when type == typeof(long) => Int64Serializer.Instance,
+            _ when type == typeof(ObjectId) => ObjectIdSerializer.Instance,
+            _ when type == typeof(TimeSpan) => new TimeSpanSerializer(),
+            _ when type == typeof(sbyte) => new SByteSerializer(),
+            _ when type == typeof(float) => new SingleSerializer(),
+            _ when type == typeof(string) => new StringSerializer(),
+            _ when type == typeof(ushort) => new UInt16Serializer(),
+            _ when type == typeof(uint) => new UInt32Serializer(),
+            _ when type == typeof(ulong) => new UInt64Serializer(),
+            _ when type == typeof(Decimal128) => new Decimal128Serializer(),
+            _ when type.IsEnum => EnumSerializer.Create(type),
+            {IsGenericType: true} when type.GetGenericTypeDefinition() == typeof(Nullable<>)
+                => CreateNullableSerializer(type.GetGenericArguments()[0]),
+            {IsGenericType: true} or {IsArray: true } => new CollectionsSerializationProvider().GetSerializer(type),
```

Comment:
nit: extra space in `{IsArray: true }`
Don't we need to check just `type is IList` for `CollectionsSerializationProvider` ?

Reply:
Will see what I can do there.

Reply:
It turns out it needs to be IList<T> which is a bit of a pain to check here in the switch so leaving as-is.